File: 02/code/objc/Lection02/Lection02/ViewController.m
//
//  ViewController.m
//  Lection02
//
//  Created by d.taraev on 16.03.17.
//  Copyright © 2017 mail.ru. All rights reserved.
//

#import "ViewController.h"
#import "Greeting.h"
#import "XIBViewController.h"
#import "HelloViewController.h"

static NSString * const segueName = @"Segue1";

@interface ViewController ()

@property (weak, nonatomic) IBOutlet UITextField *nameField;
@property (weak, nonatomic) IBOutlet UILabel *helloLabel;
@property (weak, nonatomic) IBOutlet UILabel *ageLabel;
@property (weak, nonatomic) IBOutlet UIStepper *ageStepper;
@property (weak, nonatomic) IBOutlet UISegmentedControl *genderSegment;
@property (weak, nonatomic) IBOutlet UISwitch *enabledSwitch;
@property (weak, nonatomic) IBOutlet UIView *extendedView;

@property (strong, nonatomic) Greeting *greeting;
@property (strong, nonatomic) NSString *name;

@end

@implementation ViewController

- (void)setGreeting:(Greeting *)greeting {
    
    _greeting = greeting;
    self.helloLabel.text = [greeting greeting];
}

- (NSString *)name {
    
    NSString *text = self.nameField.text ? self.nameField.text : @"";
    return text.length > 0 ? text : @"Анон";
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.greeting = [[Greeting alloc] init];
}

- (IBAction)helloButtonPressed:(id)sender {
    
    [self performSegueWithIdentifier:segueName sender:nil];
}

- (IBAction)valueChanged:(id)sender {
    
    Gender gender = self.genderSegment.selectedSegmentIndex;
    NSInteger age = self.ageStepper.value ? self.ageStepper.value : 0;
    
    self.ageLabel.text = [NSString stringWithFormat:@"Лет: %ld", (long)age];
    self.greeting = [[Greeting alloc] initWithName:self.name gender:gender age:age];
}

- (IBAction)extendedEnabledChanged:(UISwitch *)sender {
    
    BOOL isEnabled = sender.isOn;
    self.extendedView.hidden = !isEnabled;
}

- (IBAction)alphaSliderChanged:(UISlider *)sender {
    
    self.enabledSwitch.alpha = sender.value;
}

- (IBAction)showXIBPressed:(id)sender {
    XIBViewController *viewController = [[XIBViewController alloc] initWithNibName:@"XIBViewController" bundle:nil];;
    
    viewController.test = @"test";
    [self.navigationController pushViewController:viewController animated:YES];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    
    if ([segue.identifier isEqualToString:segueName]) {
        
        HelloViewController *viewController = (HelloViewController *)segue.destinationViewController;
        viewController.name = self.name;
    }
}

@end

File: 02/code/Lection02/ViewController.swift
//
//  ViewController.swift
//  Lection02
//
//  Created by Gennady Evstratov on 11/03/2017.
//  Copyright © 2017 Mail. All rights reserved.
//

import UIKit

class SegeuIdentifiers {
    static let segueName = "Segue1"
}


class ViewController: UIViewController {

    @IBOutlet weak var nameField: UITextField?
    @IBOutlet weak var helloLabel: UILabel?
    @IBOutlet weak var ageLabel: UILabel?
    @IBOutlet weak var ageStepper: UIStepper?
    @IBOutlet weak var genderSegment: UISegmentedControl?
    @IBOutlet weak var enabledSwitch: UISwitch?
    @IBOutlet weak var extendedView: UIView?
    
    fileprivate var greeting: Greeting = Greeting.empty {
        didSet {
            helloLabel?.text = greeting.greeting
        }
    }

    var name: String {
        let text = nameField?.text ?? ""
        
        return text.isEmpty ? "Анон" : text
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.

        
    }

    @IBAction func helloButtonPressed(_ sender: Any) {
        performSegue(withIdentifier: segueName, sender: nil)
    }

    @IBAction func valueChanged(_ sender: Any) {
        let gender = Gender(rawValue: genderSegment?.selectedSegmentIndex ?? 2) ?? .unknown
        let age  = Int(ageStepper?.value ?? 0)

        ageLabel?.text = "Лет: \(age)"
        greeting = Greeting(name: name, gender: gender, age: age)
    }
    
    @IBAction func extendedEnabledChanged(_ sender: Any) {
        guard let isEnabled = (sender as? UISwitch)?.isOn else {
            return
        }

        extendedView?.isHidden = !isEnabled
    }
    
    @IBAction func alphaSliderChanged(_ sender: Any) {
        guard let slider = sender as? UISlider else {
            return
        }
        
        enabledSwitch?.alpha = CGFloat(slider.value)
    }
    
    @IBAction func showXIBPressed(_ sender: Any) {
        let viewController = XIBViewController(nibName: "XIBViewController", bundle: nil)
        
        viewController.test = "test"
        
        navigationController?.pushViewController(viewController, animated: true)
    }
    
    // MARK: -
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        guard let segueIdentifier = segue.identifier else {
            return
        }
        
        switch segueIdentifier {
        case SegueIdentifiers.segueName:
            guard let viewController = segue.destination as? HelloViewController else {
                return
            }            
            viewController.name = name
            
        default:
            break
        }
    }
}


